Add listType annotations

abayer opened the pull request:
# Changes

We need to have a `// +listType=...` annotation on all list fields in API types
to get `k8s.io/kube-openapi/cmd/openapi-gen` to not throw rules violations for
list fields without that annotation. I chose to use `atomic` because it seemed
the simplest option - it means that when updates are written, the whole list
will be updated, rather than individual elements in the list, and without any
special logic on uniqueness within the list or ordering by some field in the
structs in the list.

See https://github.com/kubernetes/kube-openapi/blob/3ee0da9b0b4211c407396d9c233b38b77ce19773/pkg/idl/doc.go#L21-L63 for more information on `listType`s.

/kind bug

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [ ] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in or deleted (only if no user facing changes)

# Release Notes

Not sure what to do re: release note here. =)

<!--
Describe any user facing changes here, or delete this block.

Examples of user facing changes:
- API changes
- Bug fixes
- Any changes in behavior
- Changes requiring upgrade notices or deprecation warnings

For pull requests with a release note:

```release-note
Your release note here
```

For pull requests that require additional action from users switching to the new release, include the string "action required" (case insensitive) in the release note:

```release-note
action required: your release note here
```

For pull requests that don't need to be mentioned at release time, use the `/release-note-none` Prow command to add the `release-note-none` label to the PR. You can also write the string "NONE" as a release note in your PR description:

```release-note
NONE
```
-->

jerop commented:
looks like this change would fix https://github.com/tektoncd/pipeline/issues/3323 🤔

abayer commented:
huh! Yeah, it does look that way. Hooray for accidentally fixing additional bugs, I guess? =)

abayer commented:
/assign @jerop 
/assign @vdemeester 

Mind taking a look? I think this is good to go, and it's a v1 blocker. =)

vdemeester commented on pkg/apis/pipeline/v1beta1/pipeline_types.go:
```diff
@@ -169,11 +175,12 @@ type PipelineTask struct {
+	// When is a list of when expressions that need to be true for the task to run
 	// +optional
-	WhenExpressions WhenExpressions `json:"when,omitempty"`
+	When WhenExpressions `json:"when,omitempty"`
```
Any reason for this change ? 🙃

abayer replied:
The OpenAPI rules consider the field name not matching the JSON tag to be a rule violation, so I renamed this (and the other one) to match. If we want to get to no rule violations, we will need to rename the fields eventually.

vdemeester replied:
Lol.. really ?!? That's.. interesting.. it is in the code, it doesn't "show" to the user..

abayer replied:
Yeah, I can only assume they want to make sure the OpenAPI field name matches the Go API field name for consistency. 🤷

vdemeester commented on pkg/apis/pipeline/v1beta1/taskrun_types.go:
```diff
@@ -179,29 +185,35 @@ type TaskRunStatusFields struct {
 	// +optional
-	TaskRunResults []TaskRunResult `json:"taskResults,omitempty"`
+	// +listType=atomic
+	TaskResults []TaskRunResult `json:"taskResults,omitempty"`
```
Any reason for this change ?

vdemeester reviewed (commented):
SGTM except for the field name changes, I am not sure if they need to be on the same PR, and what is the reason behind it — especially as it might break "downstream" users of the `pkg/apis` (such as cli, …) on upgrading their dependencies.

abayer commented on hack/update-openapigen.sh:
```diff
@@ -17,15 +17,39 @@
+    --go-header-file hack/boilerplate/boilerplate.go.txt \
+    -r "${TMP_DIFFROOT}/api-report"
+
+violations=$(diff --changed-group-format='%>' --unchanged-group-format='' "hack/ignored-openapi-violations.list" "${TMP_DIFFROOT}/api-report" || echo "")
```
@vdemeester (and also @dibyom, since we'd want the same thing in Triggers once it's calling `openapi-gen`):

So this will diff the list of API violations from running `openapi-gen` against a known list of ignored violations (in `hack/ignored-openapi-violations.list`) and error out if there are any lines in in the report that aren't in the ignore list. I don't love it, but it works and allows us not to rename fields that currently violate `names_match`.

abayer replied:
@vdemeester (and also @dibyom, since we'd want the same thing in Triggers once it's calling `openapi-gen`):

So this will diff the list of API violations from running `openapi-gen` against a known list of ignored violations (in `hack/ignored-openapi-violations.list`) and error out if there are any lines in in the report that aren't in the ignore list. I don't love it, but it works and allows us not to rename fields that currently violate `names_match`.

The output will look like:

```
You appear to be running from outside of GOPATH.
This script will create a temporary GOPATH at /Users/abayer/go/src/github.com/tektoncd/pipeline/.gopath for code generation.
Moving to /Users/abayer/go/src/github.com/tektoncd/pipeline/.gopath/src/github.com/tektoncd/pipeline
Generating OpenAPI specification ...

New API rule violations found which are not present in hack/ignored-openapi-violations.list. Please fix these violations:
API rule violation: names_match,github.com/tektoncd/pipeline/pkg/apis/resource/v1alpha1,PipelineResourceSpec,SecretParams

Moving to /Users/abayer/go/src/github.com/tektoncd/pipeline
Deleting symlink to pipelines repo /Users/abayer/go/src/github.com/tektoncd/pipeline/.gopath/src/github.com/tektoncd/pipeline
When you are finished with codegen you can safely run the following:
sudo rm -rf ".gopath"
```

abayer commented:
/retest

vdemeester reviewed (approved):
/cc @tektoncd/core-maintainers

jerop commented on pkg/apis/pipeline/v1beta1/taskrun_types.go:
```diff
@@ -263,15 +275,15 @@ func (trs *TaskRunStatus) SetCondition(newCond *apis.Condition) {
 	corev1.ContainerState `json:",inline"`
 	Name                  string `json:"name,omitempty"`
-	ContainerName         string `json:"container,omitempty"`
+	Container             string `json:"container,omitempty"`
```
I believe this should also not be renamed as discussed in https://github.com/tektoncd/pipeline/pull/4402#discussion_r787932214

abayer replied:
Done.

jerop commented on pkg/apis/pipeline/v1beta1/taskrun_types.go:
```diff
@@ -304,8 +316,8 @@ type CloudEventDeliveryState struct {
 	SentAt *metav1.Time `json:"sentAt,omitempty"`
-	// Error is the text of error (if any)
-	Error string `json:"message"`
+	// Message is the text of error (if any)
+	Message string `json:"message"`
```
I believe this should also not be renamed as discussed in https://github.com/tektoncd/pipeline/pull/4402#discussion_r787932214

abayer replied:
Done.

jerop reviewed (commented):
thanks @abayer!

abayer commented on pkg/apis/pipeline/v1beta1/resource_types.go:
```diff
@@ -129,7 +134,7 @@ type PipelineResourceResult struct {
 	// See https://github.com/tektoncd/pipeline/issues/2694 for more information.
 	ResourceRef *PipelineResourceRef `json:"resourceRef,omitempty"`
-	ResultType  ResultType           `json:"type,omitempty"`
+	Type        ResultType           `json:"type,omitempty"`
```
I think it (and the other ones you commented on) should be changed - otherwise, we have to homebrew some way to ignore these specific cases of the field name not matching the JSON tag.

abayer commented:
Ok, rebased, additional rules violations which had been added since I last rebased are addressed (including a number of fields which didn't have `json` tags at all), and four fields have been un-renamed and their resulting violations in the output are now ignored.

abayer commented:
/retest

abayer commented:
/retest

abayer commented:
/retest

abayer commented:
/retest

abayer commented:
/retest

abayer commented on pkg/apis/pipeline/v1beta1/taskrun_types.go:
```diff
@@ -263,15 +275,15 @@ func (trs *TaskRunStatus) SetCondition(newCond *apis.Condition) {
 	corev1.ContainerState `json:",inline"`
 	Name                  string `json:"name,omitempty"`
-	ContainerName         string `json:"container,omitempty"`
+	Container             string `json:"container,omitempty"`
```
Done

lbernick replied:
this still appears to be renamed?

abayer replied:
Ha, I renamed the wrong field entirely when "reverting". 🤦

lbernick commented on pkg/apis/pipeline/v1beta1/resolver_types.go:
```diff
@@ -33,15 +33,16 @@ type ResolverRef struct {
 	// Name is the name of the parameter that will be passed to the
 	// resolver.
-	Name string
+	Name string `json:"name,omitempty"`
```
for fields with "json:omitempty" added, should they have a "+optional" comment as well?

abayer replied:
Good question - in these cases, they probably shouldn't be `omitempty` in the first place. I'll clean that up.

lbernick commented on pkg/apis/pipeline/v1beta1/resultref.go:
```diff
@@ -23,6 +23,7 @@ import (
 )
 
 // ResultRef is a type that represents a reference to a task run result
+// +k8s:openapi-gen=false
```
what's the reason for this?

abayer replied:
`ResultRef` isn't actually used anywhere, so far as I can tell. Being me with my tendency to not muck around with things I don't actually have full context on, I decided that it didn't make sense to care about whether `ResultRef`'s fields were properly tagged if it's never used but didn't want to go too far and delete the type completely. So I opted to exclude `ResultRef` from the OpenAPI stuff entirely, hence the annotation.

lbernick replied:
it looks like it's used here https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/resources/resultrefresolution.go so I would include it in codegen

lbernick replied:
It doesn't look like this comment was addressed

abayer replied:
Ah, don't know how I missed both the usage and your comment! 🤦 Fixing now!

lbernick reviewed (commented):
Would you mind adding a bit more context to the commit message and PR description about what this annotation does and why it is added? Having detail in the commit message is useful when people are using tools like git log, to be able to understand more about why the change was made without having to context switch by reading through the linked issue.

abayer commented:
@lbernick - I've removed the probably-not-appropriate `omitempty`s and added more detail to the commit message. I've left `ResultRef` alone at least for now.

abayer commented:
/retest

lbernick reviewed (commented):
Thanks Andrew! Just one last quick comment, could you please update the commit message/PR title to reflect that fields are not being renamed, and link to list type [docs](https://github.com/kubernetes/kube-openapi/blob/3ee0da9b0b4211c407396d9c233b38b77ce19773/pkg/idl/doc.go#L21-L63)?

lbernick commented:
/lgtm
